src/Microsoft.Dnx.Runtime/DependencyManagement/CompatibilityChecker.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.Versioning;
+using Microsoft.Dnx.Runtime.DependencyManagement;
+using NuGet;
+
+namespace Microsoft.Dnx.Runtime
+{
+    public class CompatibilityChecker
+    {
+        private readonly LockFile _lockFile;
+        private readonly ILookup<string, LockFileLibrary> _lockFileLibraries;
+        private readonly Dictionary<FrameworkName, HashSet<string>> _frameworkRuntimeAssemblies;
+
+        public CompatibilityChecker(LockFile lockFile)
+            : this(new OptimizedLockFile(lockFile))
+        {
+        }
+
+        public CompatibilityChecker(OptimizedLockFile optimizedLockFile)
+        {
+            _lockFile = optimizedLockFile.LockFile;
+            _lockFileLibraries = optimizedLockFile.LockFileLibraryLookup;
+            _frameworkRuntimeAssemblies = _lockFile.Targets.ToDictionary(
+                keySelector: t => t.TargetFramework,
+                elementSelector: GetRuntimeAssemblySet);
+        }
+
+        public IEnumerable<CompatibilityIssue> GetAllCompatibilityIssues()
+        {
+            foreach (var target in _lockFile.Targets)
+            {
+                foreach (var library in target.Libraries)
+                {
+                    var issue = CheckTargetLibrary(library, target.TargetFramework);
+                    if (issue != null)
+                    {
+                        yield return issue;
+                    }
+                }
+            }
+        }
+
+        public CompatibilityIssue CheckTargetLibrary(LockFileTargetLibrary targetLibrary, FrameworkName framework)
+        {
+            var lockFileLibrary = _lockFileLibraries[targetLibrary.Name].First(l => l.Version == targetLibrary.Version);
```

Comment:
Because I wanted to save memory here by having a reusable lookup in `OptimizedLockFile`. If we make the lookup based on Tuple(id, version), then `PackageRepository.FindPackagesById()` cannot reuse this. Need to do profiling to make the tradeoff here.